### PR TITLE
bpo-37394: Fix pure Python implementation of queue

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue37394

### CODE REVIEW

**Comment by vstinner** [comment]
test_queue didn't catch this issue, so test_queue should be fixed somehow.

IMHO test_queue should be rewritten using something like that:

```
py_functools = support.import_fresh_module('functools', blocked=['_functools'])
c_functools = support.import_fresh_module('functools', fresh=['_functools'])
```

File: Lib/test/test_queue.py [hunk 1]
```diff
@@ -232,45 +228,88 @@ def test_shrinking_queue(self):
...
+
+    def setUp(self):
+        self.type2test = self.queue.Queue
+        super().setUp()
+
+class PyQueueTest(QueueTest, unittest.TestCase):
+    queue = py_queue
+
+
+@unittest.skipUnless(c_queue, "No _queue module found")
```
Comment (vstinner) [inline Lib/test/test_queue.py]: Maybe you can use:

```
need_c_queue = unittest.skipUnless(c_queue, "No _queue module found")

@need_c_queue
class CQueueTest(QueueTest, unittest.TestCase):  ...
```

File: Lib/test/test_queue.py [hunk 2]
```diff
@@ -232,45 +228,88 @@ def test_shrinking_queue(self):
...
-        return queue.Queue._get(self)
-
-class FailingQueueTest(BlockingTestMixin, unittest.TestCase):
+class FailingQueueException(Exception): pass
+
+class FailingQueueTest(BlockingTestMixin):
+
+    def setUp(self):
+
+        queue_module = self.queue
```
Comment (vstinner) [inline Lib/test/test_queue.py]: nitpick: maybe you can use there: Queue = self.queue.Queue

Comment (pablogsal) [reply Lib/test/test_queue.py]: Done in 
2175091

File: Lib/test/test_queue.py (hunk 1 above)
Comment (pablogsal) [reply Lib/test/test_queue.py]: Done in 2175091

**Review by vstinner — APPROVED** [review]
LGTM. I suggest to rephrase the commit title, mayby: "Fix pure Python implementation of queue".

I may be interesting to add NEWS entry, and maybe backport this change to 3.7 and 3.8. You have to open a bpo issue in this case (to get a bpo number).

**Pull request merged**